File: 704. Binary Search/solution.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;
        int pivot;
        while (left <= right) {
            pivot = left + (right - left) / 2;
            if (nums[pivot] == target) return pivot;
            if (target < nums[pivot]) right = pivot - 1;
            else left = pivot + 1;
        }
        return -1;
    }
};

void print(std::vector <int> const& a) {
    std::cout << "The answer is: ";

    for (int i = 0; i < a.size(); i++)
        std::cout << a.at(i) << ' ';
}

int main()
{
    vector<int> nums;
    int target = 9;

    nums.push_back(-1);
    nums.push_back(0);
    nums.push_back(3);
    nums.push_back(5);
    nums.push_back(9);
    nums.push_back(12);

    Solution solution;

    std::cout << "The answer is: ";
    std::cout << solution.search(nums, target);
 
    return 0;
}

File: 704. Binary Search/solution.java

class Solution {
    public int search(int[] nums, int target) {
    	int left = 0;
    	int right = nums.length-1;
    	int pivot;
    	while (left<= right) {
    		pivot = left + (right - left) / 2;
    	    if (nums[pivot] == target) return pivot;
            if (target < nums[pivot]) right = pivot - 1;
            else left = pivot + 1;
    	}
        // In case there is no solution, we'll just return -1
        return -1;
    }
}


public class main {
	public static void main(String[] args) {
		Solution solution = new Solution();
		int[] nums;
		nums = new int[6];
		nums[0] = -1;
		nums[1] = 0;
		nums[2] = 3;
		nums[3] = 5;
		nums[4] = 9;
		nums[5] = 15;
		int target = 9;
	    System.out.println("The answer is:");
	    System.out.println(solution.search(nums, target));
	}
}
